crates/btcio/src/writer/builder.rs

```diff
@@ -494,26 +489,28 @@ fn assert_correct_address(
 
 // Generates a [`ScriptBuf`] that consists of `OP_IF .. OP_ENDIF` block
 pub fn generate_inscription_script(
-    inscription_data: InscriptionData,
+    inscriptions: &[InscriptionBlob],
     rollup_name: &str,
-    version: u8,
 ) -> anyhow::Result<ScriptBuf> {
-    let mut builder = script::Builder::new()
-        .push_opcode(OP_FALSE)
-        .push_opcode(OP_IF)
-        .push_slice(PushBytesBuf::try_from(ROLLUP_NAME_TAG.to_vec())?)
-        .push_slice(PushBytesBuf::try_from(rollup_name.as_bytes().to_vec())?)
-        .push_slice(PushBytesBuf::try_from(VERSION_TAG.to_vec())?)
-        .push_slice(PushBytesBuf::from([version]))
-        .push_slice(PushBytesBuf::try_from(BATCH_DATA_TAG.to_vec())?)
-        .push_int(inscription_data.batch_data().len() as i64);
-
-    trace!(batchdata_size = %inscription_data.batch_data().len(), "Inserting batch data");
-    for chunk in inscription_data.batch_data().chunks(520) {
-        trace!(size=%chunk.len(), "inserting chunk");
-        builder = builder.push_slice(PushBytesBuf::try_from(chunk.to_vec())?);
+    let mut builder = script::Builder::new();
+    for (index, inscription) in inscriptions.iter().enumerate() {
+        builder = builder.push_opcode(OP_FALSE).push_opcode(OP_IF);
+
+        if index == 0 {
+            builder = builder.push_slice(PushBytesBuf::try_from(rollup_name.as_bytes().to_vec())?);
+        }
+
+        builder = builder
+            .push_int(inscription.data_type() as i64) // Batch Tag
+            .push_int(inscription.data().len() as i64);
+
+        trace!(batchdata_size = %inscription.data().len(), "Inserting batch data");
+        for chunk in inscription.data().chunks(520) {
+            trace!(size=%chunk.len(), "inserting chunk");
+            builder = builder.push_slice(PushBytesBuf::try_from(chunk.to_vec())?);
+        }
+        builder = builder.push_opcode(OP_ENDIF);
     }
```

Comment:
This needs to be reworked some more to reflect the format as written in the ticket description.

```
OP_FALSE
OP_IF
  PUSH(tag)
  PUSH(payload[..512])
  PUSH(payload[512..1024])
  ...
OP_ENDIF
```

We're using the type tags also as the magic numbers, so it doesn't need to take `rollup_name` anymore, but instead it should be using the tag strings as resolved from the `InscriptionBlob` type.  This pushes an int based on the variant, which is also implicitly making it dependent on the order that they're listed in the enum.

Perhaps we should lower it down again, resolve the tag bytes from the params based on the blob types ahead of time, and have a function that takes blobs with explicit tag bytes as slices instead?

I'm fairly sure we don't need to also push the length of the payload there do we?  It's implicit based on the number of pushes after that until the ENDIF.

Reply:
> instead it should be using the tag strings as resolved from the InscriptionBlob type.
I still think, `rollup_name` makes more sense than the different tag strings. We just have to include  the rollup name for the first envelope then it can be used for filtering. Later included envelopes can just include a integer based tag which is byte efficient than posting the tag string every time.

Reply:
(discussed in DMs)

Ideally we should want to avoid special casing the parsing and having to treat all of the envelopes together.  Because then they're not logically distinct envelopes, they're a single data structure.  Also raises the question of there being envelopes in different inputs, do we key those on the field from a different envelope on a different input?